src/plugins/FingerPrint.js

```diff
@@ -0,0 +1,395 @@
+        'SCAN_INSUFFICIENT': 'scan insufficient',
+        'SET_AUTO_RECOGNIZE_FALSE': 'set auto_recognize false',
+        'SET_AUTO_RECOGNIZE_TRUE': 'set auto_recognize true',
+        'NOTIFY_ALL': 'notify all',
+    },
+    toReceive: {
+        'SCAN_ERROR': 'scan error',
+        'SCAN_ERROR_TIMEOUT': 'scan error timeout',
+        'SCAN_ERROR_CANCELED': 'scan error canceled',
+        'CURRENT_STATUS_IDLE': 'current_status idle',
+        'CURRENT_STATUS_SCANNING': 'current_status scanning',
+        'CURRENT_STATUS_ENROLLING': 'current_status enrolling',
+        'SCAN_START': 'scan start',
+        'SCAN_SUCCESS': 'scan success',
+        'AUTO_RECOGNIZE_FALSE': 'auto_recognize false',
+        'AUTO_RECOGNIZE_TRUE': 'auto_recognize true',
+    }
+};
+module.exports = class FingerPrint extends OverlayPlugin {
+    /**
+     * Plugin initialization.
+     *
+     * @param {Object} instance Associated instance.
+     * @param {Object} i18n     Translations keys for the UI.
+     */
+    constructor(instance, i18n) {
+        super(instance);
+
+        // Reference instance
+        this.instance = instance;
+        this.i18n = i18n || {};
+
+        // Register plugin
+        this.instance.fingerprint = this;
+
+        // proxy for unidirectionnal modification (state -> widget)
+        this.state = new Proxy({
+            isScanStart: false,
+            isRecognizedFPByDefault: false,
+            isStatusIdle: false,
+            isEnrolling: false,
+            isScanning: false,
+        }, {
+            set: (state, prop, value) => {
+                const oldValue = state[prop];
+                state[prop] = value;
+
+                // post update process
+                if ((state.isEnrolling || state.isScanning) && !state.isRecognizedFPByDefault) {
+                    this.enableBody();
+                } else {
+                    this.disableBody();
+                }
+
+                const authRequiredDiv = document.querySelector('.gm-fingerprint-dialog-auth-required-status');
+
+                switch (prop) {
+                case 'isStatusIdle':
+                    break;
+                case 'isScanStart':
+                case 'isEnrolling':
+                case 'isScanning':
+                    if (state.isScanStart && (state.isEnrolling || state.isScanning)) {
+                        authRequiredDiv.classList.add('active');
+                        authRequiredDiv.setAttribute('data-text', this.i18n.yes ?? 'Yes');
+                        this.toolbarBtnImage.classList.add('fingerprint-require');
+                    } else {
+                        authRequiredDiv.classList.remove('active');
+                        authRequiredDiv.setAttribute('data-text', this.i18n.no ??  'No');
+                        this.toolbarBtnImage.classList.remove('fingerprint-require');
+                    }
+                    break;
+                case 'isRecognizedFPByDefault':
+                    if (value) {
```

Comment:
it doesn't produce the same result. This piece of code is maybe not really cleaner. 

` 
case FINGERPRINT_MESSAGES.toReceive.AUTO_RECOGNIZE_FALSE:
      this.state.isRecognizedFPByDefault = false;
      break;
  case FINGERPRINT_MESSAGES.toReceive.AUTO_RECOGNIZE_TRUE:
      this.state.isRecognizedFPByDefault = true;
      break;
`
....
`
 const recognizedFPByDefaultStatus = switchButton.createSwitch({
      onChange: (value) => {
          this.state.isRecognizedFPByDefault = value;
      },
  });
`
 
The probelm here is that ` this.state.isRecognizedFPByDefault = value;` will change isRecognizedFPByDefault, so an event `SET_AUTO_RECOGNIZE_BOOL` is send. The HAL fire an `FINGERPRINT_MESSAGES.toReceive.AUTO_RECOGNIZE_BOOL` event and we got an infinite loop. To avoid this i have conditionned the trigger of event with

`
 if (value) {
   !oldValue && sendEvent
   SYNC_UI
`

and 

`
 if (value) {
   !oldValue && sendEvent
   SYNC_UI
`

But for somes scenario we need to sync the UI even if internal state is right

I can write 

`
case 'isRecognizedFPByDefault':
      if (value !== oldValue) {
          if (value) {
             this.sendDataToInstance(FINGERPRINT_MESSAGES.toSend.SET_AUTO_RECOGNIZE_TRUE);
          } else {
             this.sendDataToInstance(FINGERPRINT_MESSAGES.toSend.SET_AUTO_RECOGNIZE_FALSE);
          }
      }
      if (value) {
           this.toolbarBtnImage.classList.add('fingerprint-autoValidation');        
      } else {
          this.toolbarBtnImage.classList.remove('fingerprint-autoValidation');
      }
      //update switch
      document.querySelector('.gm-fingerprint-dialog-recognized-fp-by-default-status').setState(value);
      break;
`

if it's cleaner for you

Reply:
You proposal looks good to me. Maybe add a comment on why we still need to update the UI even if the value does not change.